### ISSUE TITLE
[Relay] authentication encountered error: could not send request to upstream

### ISSUE THREAD

**Issue by swcleo** [sentry-kubernetes/charts#175]
when I use helm to deploy sentry on my k8s cluster, Pod relay is not ready and get some error.

There is a paragraph in README

As Relay is now part of this chart your need to make sure you enable either Nginx or the Ingress. Please read the next paragraph for more informations.

If you are using an ingress gateway (like istio), you have to change your inbound path from sentry-web to nginx.

**Can anyone tell me what's wrong**

```
2020-09-17T06:03:16Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T06:03:16Z [relay_server::actors::upstream] ERROR: authentication encountered error: upstream request returned error 500 Internal Server Error2020-09-17T08:43:10Z [rdkafka::client] ERROR: librdkafka: Global error: BrokerTransportFailure (Local: Broker transport failure): sentry-kafka:9092/bootstrap: Connect to ipv4#10.103.165.17:9092 failed: Connection refused (after 1023ms in state CONNECT)
2020-09-17T08:43:10Z [rdkafka::client] ERROR: librdkafka: Global error: AllBrokersDown (Local: All broker connections are down): 1/1 brokers are down
2020-09-17T08:43:12Z [relay_server::actors::upstream] ERROR: authentication encountered error: could not send request to upstream
  caused by: Timeout while waiting for response
2020-09-17T08:43:12Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T08:43:15Z [relay_server::actors::upstream] ERROR: authentication encountered error: could not send request to upstream
  caused by: Timeout while waiting for response
2020-09-17T08:43:16Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T08:43:19Z [relay_server::actors::upstream] ERROR: authentication encountered error: could not send request to upstream
  caused by: Timeout while waiting for response
2020-09-17T08:43:21Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T08:43:22Z [relay_server::actors::upstream] ERROR: authentication encountered error: could not send request to upstream
  caused by: Failed to connect to host: Connection refused (os error 111)
  caused by: Connection refused (os error 111)
  caused by: Connection refused (os error 111)
2020-09-17T08:43:24Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T08:43:25Z [relay_server::actors::upstream] ERROR: authentication encountered error: upstream request returned error 500 Internal Server Error
  caused by: Internal Error
2020-09-17T08:43:29Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T08:43:29Z [relay_server::actors::upstream] ERROR: authentication encountered error: upstream request returned error 500 Internal Server Error
  caused by: Internal Error
2020-09-17T08:43:34Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T08:43:34Z [relay_server::actors::upstream] ERROR: authentication encountered error: upstream request returned error 500 Internal Server Error
  caused by: Internal Error
2020-09-17T08:43:41Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T08:43:41Z [relay_server::actors::upstream] ERROR: authentication encountered error: upstream request returned error 500 Internal Server Error
  caused by: Internal Error
2020-09-17T08:43:53Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T08:43:53Z [relay_server::actors::upstream] ERROR: authentication encountered error: upstream request returned error 500 Internal Server Error
  caused by: Internal Error
2020-09-17T08:44:10Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T08:44:10Z [relay_server::actors::upstream] ERROR: authentication encountered error: upstream request returned error 500 Internal Server Error
  caused by: Internal Error
2020-09-17T08:44:36Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T08:44:36Z [relay_server::actors::upstream] ERROR: authentication encountered error: upstream request returned error 500 Internal Server Error
  caused by: Internal Error
2020-09-17T08:45:14Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T08:45:14Z [relay_server::actors::upstream] ERROR: authentication encountered error: upstream request returned error 500 Internal Server Error
  caused by: Internal Error
2020-09-17T08:46:12Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-17T08:46:12Z [relay_server::actors::upstream] ERROR: authentication encountered error: could not send request to upstream
  caused by: Server disconnected
  caused by: Server disconnected
```

**Comment by Mokto** [comment]
It's a known issue : the first time you install the chart, it takes some time to register relay properly.

When the pod restarts it should be fine ?

**Comment by swcleo** [comment]
I tried to wait for 20 minutes to check the execution status, but pod relay keep restarting all the time.
This is my helm chart, it can be used to simulate the situation.

https://github.com/swcleo/sentry-test. 

```
helm install sentry -n sentry . --wait -f test.yaml
```

**Comment by Mokto** [comment]
Out of curiosity, have you done the UI setup on your sentry installation ?

**Comment by swcleo** [comment]
> Out of curiosity, have you done the UI setup on your sentry installation ?

No.

**Comment by swcleo** [comment]
> Out of curiosity, have you done the UI setup on your sentry installation ?

no. It is currently still in the helm install stage, and even the job that created the sentry-db-init has not been started yet

**Comment by swcleo** [comment]
> It's a known issue : the first time you install the chart, it takes some time to register relay properly.
> 
> When the pod restarts it should be fine ?

May I ask how long is the "some time"?

**Comment by swcleo** [comment]
03:54:35 [WARNING] django.security.csrf: Forbidden (CSRF cookie not set.): /api/1/store/ (status_code=403 request=<WSGIRequest: POST u'/api/1/store/?sentry_key=69b8a633500f449fb7766e35378ad46c&sentry_version=7'>)
10.244.0.5 - - [18/Sep/2020:03:54:35 +0000] "POST /api/1/store/?sentry_key=69b8a633500f449fb7766e35378ad46c&sentry_version=7 HTTP/1.1" 403 5027 "http://localhost:3000/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.102 Safari/537.36"

**Comment by michaelquong** [comment]
I am also getting a similar problem.
I have had tagged version 4.2.0 running prior but wanted to upgrade to the latest tagged version 5.1.0

For myself, doing a helm upgrade all components came up except for relay.

**Comment by WolfspiritM** [comment]
EDIT:
**I think I found the problem. If you run the chart with --wait or --atomic then the post install hooks are run as soon as everything is done. However without that post install scripts the chart never gets done cause the pods don't come up cause of missing migrations.** As soon as I removed --atomic the chart succeeds and pods come up.

----


Similar issue here. I'm trying to install this chart for hours now without any success.
For me it's not only the relay but also the web that is crashing.
```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/sentry/api/base.py", line 134, in handle_exception
    response = super(Endpoint, self).handle_exception(exc)
  File "/usr/local/lib/python2.7/site-packages/rest_framework/views.py", line 449, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python2.7/site-packages/sentry/api/base.py", line 247, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/sentry/api/endpoints/relay_register.py", line 115, in post
    relay = Relay.objects.get(relay_id=relay_id)
  File "/usr/local/lib/python2.7/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/django/db/models/query.py", line 374, in get
    num = len(clone)
  File "/usr/local/lib/python2.7/site-packages/django/db/models/query.py", line 232, in __len__
    self._fetch_all()
  File "/usr/local/lib/python2.7/site-packages/django/db/models/query.py", line 1121, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/usr/local/lib/python2.7/site-packages/django/db/models/query.py", line 53, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch)
  File "/usr/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 899, in execute_sql
    raise original_exception
ProgrammingError: UndefinedTable('relation "sentry_relay" does not exist\nLINE 1: ...y"."last_seen", "sentry_relay"."is_internal" FROM "sentry_re...\n                                                             ^\n',)
SQL: SELECT "sentry_relay"."id", "sentry_relay"."relay_id", "sentry_relay"."public_key", "sentry_relay"."first_seen", "sentry_relay"."last_seen", "sentry_relay"."is_internal" FROM "sentry_relay" WHERE "sentry_relay"."relay_id" = %s
```

At some time the snuba services crash for me aswell.
```Subscribed topic not available: event-replacements: Broker: Unknown topic or partition```

It feels like there is some kind of chicken-and-egg problem. As if the migrations aren't run cause the services doing the migrations is waiting for another service to come up that needs the database already to answer. Or maybe migrations aren't run at all somehow.

**Comment by swcleo** [comment]
Relay will wait until the database is created. 
I try to comment out `deployment-relay.yaml` "livenessProbe" and "readinessProbe", complete all initialization jobs and all pods be Running.

When I try to test sentry, I got another problem.

```
POST http://sentry.xxx.xxx/api/2/store/?sentry_key=cbbfe69801e74fd48b6f9d192e939bed&sentry_version=7 403 (Forbidden)
```

kubectl logs -f sentry-web-xxx

```
02:03:50 [WARNING] django.security.csrf: Forbidden (CSRF cookie not set.): /api/2/store/ (status_code=403 request=<WSGIRequest: POST u'/api/2/store/?sentry_key=cbbfe69801e74fd48b6f9d192e939bed&sentry_version=7'>)
10.244.0.90 - - [21/Sep/2020:02:03:50 +0000] "POST /api/2/store/?sentry_key=cbbfe69801e74fd48b6f9d192e939bed&sentry_version=7 HTTP/1.1" 403 5027 "http://localhost:3000/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.102 Safari/537.36"
```

I am confirming whether the gateway does not match.

Ingress: Istio

```
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: sentry
  namespace: sentry
spec:
  hosts:
  - "sentry.xxx.xxx"
  gateways:
  - sentry-gateway
  http:
  - match:
    - uri:
        prefix: /
    route:
    - destination:
        port:
          number: 9000
        host: sentry-web
    corsPolicy:
      allowOrigins:
        - exact: "*"
```

**Comment by swcleo** [comment]
Reference this issue

https://github.com/sentry-kubernetes/charts/issues/153

I change virtualservice config, 

```
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: sentry
  namespace: sentry
spec:
  hosts:
  - "sentry.front.rde"
  gateways:
  - sentry-gateway
  http:
  - match:
    - uri:
        prefix: /
    route:
    - destination:
        port:
          number: 80
        host: sentry-nginx
    corsPolicy:
      allowOrigins:
        - exact: "*"
```

sentry-nginx is used to distribute the request to sentry-relay and sentry-web.

When client POST Error to sertry,

client -> nginx => Good. 

nginx -> relay => Good. 

relay -> kafka => Bad!!!  



```
2020-09-21T03:17:32Z [relay::setup] INFO: launching relay from config folder /work/.relay
2020-09-21T03:17:32Z [relay::setup] INFO:   relay mode: managed
2020-09-21T03:17:32Z [relay::setup] INFO:   relay id: ede2cfe3-d985-4d2d-b4d7-c4bc84de74f9
2020-09-21T03:17:32Z [relay::setup] INFO:   public key: pF1rcR5kDC6Hgpv0NtwZDBMue90tFT6CWjYBzMgrwT8
2020-09-21T03:17:32Z [relay::setup] INFO:   log level: INFO
2020-09-21T03:17:32Z [relay_server::actors::upstream] INFO: upstream relay started
2020-09-21T03:17:32Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-21T03:17:32Z [relay_server::actors::outcome::processing] INFO: OutcomeProducer started.
2020-09-21T03:17:33Z [r2d2] ERROR: Connection refused (os error 111)
2020-09-21T03:17:33Z [r2d2] ERROR: Connection refused (os error 111)
2020-09-21T03:17:34Z [r2d2] ERROR: Connection refused (os error 111)
2020-09-21T03:17:35Z [rdkafka::client] ERROR: librdkafka: Global error: BrokerTransportFailure (Local: Broker transport failure): sentry-kafka:9092/bootstrap: Connect to ipv4#10.98.58.86:9092 failed: Connection refused (after 3038ms in state CONNECT)
2020-09-21T03:17:35Z [rdkafka::client] ERROR: librdkafka: Global error: AllBrokersDown (Local: All broker connections are down): 1/1 brokers are down
2020-09-21T03:17:36Z [r2d2] ERROR: Connection refused (os error 111)
2020-09-21T03:17:39Z [relay_server::actors::events] INFO: starting 8 event processing workers
2020-09-21T03:17:39Z [relay_server::actors::store] INFO: store forwarder started
2020-09-21T03:17:39Z [relay_server::actors::project_redis] INFO: redis project cache started
2020-09-21T03:17:39Z [relay_server::actors::project_redis] INFO: redis project cache started
2020-09-21T03:17:39Z [relay_server::actors::project_redis] INFO: redis project cache started
2020-09-21T03:17:39Z [relay_server::actors::project_redis] INFO: redis project cache started
2020-09-21T03:17:39Z [relay_server::actors::project_redis] INFO: redis project cache started
2020-09-21T03:17:39Z [relay_server::actors::project_redis] INFO: redis project cache started
2020-09-21T03:17:39Z [relay_server::service] INFO: spawning http server
2020-09-21T03:17:39Z [relay_server::service] INFO:   listening on: http://0.0.0.0:3000/
2020-09-21T03:17:39Z [relay_server::actors::project_redis] INFO: redis project cache started
2020-09-21T03:17:39Z [actix_net::server::server] INFO: Starting 8 workers
2020-09-21T03:17:39Z [relay_server::actors::project_redis] INFO: redis project cache started
2020-09-21T03:17:40Z [actix_net::server::server] INFO: Starting server on 0.0.0.0:3000
2020-09-21T03:17:40Z [relay_server::actors::controller] INFO: relay server starting
2020-09-21T03:17:40Z [relay_server::actors::connector] INFO: metered connector started
2020-09-21T03:17:40Z [relay_server::actors::events] INFO: event manager started
2020-09-21T03:17:40Z [relay_server::actors::project_local] INFO: project local cache started
2020-09-21T03:17:40Z [relay_server::actors::project_upstream] INFO: project upstream cache started
2020-09-21T03:17:40Z [relay_server::actors::project_cache] INFO: project cache started
2020-09-21T03:17:40Z [relay_server::actors::project_keys] INFO: project cache started
2020-09-21T03:17:40Z [relay_server::actors::relays] INFO: key cache started
2020-09-21T03:17:41Z [rdkafka::client] ERROR: librdkafka: Global error: BrokerTransportFailure (Local: Broker transport failure): sentry-kafka:9092/bootstrap: Connect to ipv4#10.98.58.86:9092 failed: Connection refused (after 1023ms in state CONNECT)
2020-09-21T03:17:41Z [rdkafka::client] ERROR: librdkafka: Global error: AllBrokersDown (Local: All broker connections are down): 1/1 brokers are down
2020-09-21T03:17:43Z [relay_server::actors::upstream] ERROR: authentication encountered error: could not send request to upstream
  caused by: Timeout while waiting for response
2020-09-21T03:17:43Z [relay_server::actors::upstream] INFO: registering with upstream (http://sentry-web:9000/)
2020-09-21T03:17:46Z [relay_server::actors::upstream] ERROR: authentication encountered error: could not send request to upstream
  caused by: Timeout while waiting for response
```

**Comment by swcleo** [comment]
Finally, I finished deploy sentry service with k8s.

1. kafka replicaCount must match your config.  
 => check kafka logs

2. If relay always get ERROR: authentication encounteredm, delete current relay pod   
=> kubectl delete pod sentry-relay-xxx

**Comment by WolfspiritM** [comment]
@swcleo I still think that the chart waiting for the relay to finish before starting the migrations is a bug and should be fixed. That's why I think this issue should stay open.

**Comment by swcleo** [comment]
> @swcleo I still think that the chart waiting for the relay to finish before starting the migrations is a bug and should be fixed. That's why I think this issue should stay open.

Just like you said, it’s a “Chicken-Egg” problem. I just temporarily avoid the problem of dependency through some methods.

**Comment by Mokto** [comment]
Could you please try the code from this PR ?

https://github.com/sentry-kubernetes/charts/pull/179

I put relay as a helm hook to fix the situation. It worked on a new install for me.

**Comment by Mokto** [comment]
Fixed through 5.2.1
